app/src/test/java/org/mozilla/fenix/exceptions/login/LoginExceptionsInteractorTest.kt

```diff
@@ -6,39 +6,34 @@ package org.mozilla.fenix.exceptions.login
 
 import io.mockk.mockk
 import io.mockk.verify
-import kotlinx.coroutines.test.TestCoroutineScope
-import kotlinx.coroutines.test.runBlockingTest
+import kotlinx.coroutines.test.TestScope
+import kotlinx.coroutines.test.UnconfinedTestDispatcher
+import kotlinx.coroutines.test.runTest
 import mozilla.components.feature.logins.exceptions.LoginException
 import mozilla.components.feature.logins.exceptions.LoginExceptionStorage
-import org.junit.After
 import org.junit.Before
 import org.junit.Test
 
 class LoginExceptionsInteractorTest {
 
     private lateinit var loginExceptionStorage: LoginExceptionStorage
     private lateinit var interactor: LoginExceptionsInteractor
-    private val scope = TestCoroutineScope()
+    private val scope = TestScope(UnconfinedTestDispatcher())
 
     @Before
     fun setup() {
         loginExceptionStorage = mockk(relaxed = true)
         interactor = DefaultLoginExceptionsInteractor(scope, loginExceptionStorage)
     }
 
-    @After
-    fun cleanUp() {
-        scope.cleanupTestCoroutines()
-    }
-
     @Test
-    fun onDeleteAll() = scope.runBlockingTest {
+    fun onDeleteAll() = scope.runTest {
```

Comment:
Just out of curiosity, why are `UnconfinedTestDispatcher` and the extension function version of `runTest` required for this test class?

Reply:
There are two dispatchers to choose from: `UnconfinedTestDispatcher` and `StandardTestDispatcher`.
`UnconfinedTestDispatcher` has the nice advantage (here) that it eagerly enters `launch` and `async` blocks.
We need this since the interactor executes for example
```
ioScope.launch {
    loginExceptionStorage.removeLoginException(item)
}
```
Using the same scope (and dispatcher) in the interactor and for running the test allows to sync the execution.

Reply:
Yes, switching to `UnconfinedTestDispatcher` does make the nested coroutines launch, but they can launch on a different thread and we're not synchronizing the execution i.e., the test code is not waiting for their completion. For many tests this will not be good enough and lead to intermittent failures. I've added and example here: https://github.com/mozilla-mobile/android-components/pull/12006#pullrequestreview-960776443

It's easy to fix, but we should review the tests that launch nested coroutines and make assertion based on them.